pbsmetrics/metrics.go

```diff
@@ -0,0 +1,257 @@
+package pbsmetrics
+
+import (
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/rcrowley/go-metrics"
+	"github.com/vrischmann/go-metrics-influxdb"
+	"strings"
+	"time"
+)
+
+// Labels defines the labels that can be attached to the metrics.
+type Labels struct {
+	Source        DemandSource
+	RType         RequestType
+	Adapter       openrtb_ext.BidderName
+	PubID         string // exchange specific ID, so we cannot compile in values
+	Browser       Browser
+	CookieFlag    CookieFlag
+	RequestStatus RequestStatus
+}
+
+// Label typecasting. Se below the type definitions for possible values
+
+// DemandSource : Demand source enumeration
+type DemandSource string
+
+// RequestType : Request type enumeration
+type RequestType string
+
+// Browser type enumeration
+type Browser string
+
+// CookieFlag : User ID cookie exists flag
+type CookieFlag string
+
+// RequestStatus : The request/adapter return status
+type RequestStatus string
+
+// The demand sources
+const (
+	DemandWeb     DemandSource = "web"
+	DemandApp     DemandSource = "app"
+	DemandUnknown DemandSource = "unknown"
+)
+
+// The request types (endpoints)
+const (
+	ReqTypeLegacy RequestType = "legacy"
+	ReqTypeORTB2  RequestType = "openrtb2"
+	ReqTypeAMP    RequestType = "amp"
+)
+
+// Browser flag; at this point we only care about identifying Safari
+const (
+	BrowserSafari Browser = "safari"
+	BrowserOther  Browser = "other"
+)
+
+// Cookie flag
+const (
+	CookieFlagYes     CookieFlag = "2"
+	CookieFlagNo      CookieFlag = "1"
+	CookieFlagUnknown CookieFlag = "0"
+)
+
+// Request/return status
+const (
+	RequestStatusOK      RequestStatus = "ok"
+	RequestStatusErr     RequestStatus = "err"
+	RequestStatusNoBid   RequestStatus = "nobid"   // Only for adapters
+	RequestStatusTimeout RequestStatus = "timeout" // Only for adapters
+)
+
+// UserLabels : Labels for /setuid endpoint
+type UserLabels struct {
+	Action RequestAction
+	Bidder openrtb_ext.BidderName
+}
+
+// RequestAction : The setuid request result
+type RequestAction string
```

Comment:
Maybe an opinion, but... do you think these work well as labels, or separate methods on the interface?

With the exception of this one cluster of constants, it seems to me like all the `MetricsEngine` methods are different Events, and all the `Labels` are little packages of cross-cutting concerns.

I don't think it's a big deal either way... just wanted to point it out cause I find it a _bit_ more intuitive that way, so I'm fishing to see if you shared the uncertainty.

Reply:
I do believe these are all the same incoming event (`/setuid` request), but just enumerate how PBS responds to the request. We can refuse to do anything because of opt out, we can error on an invalid request, or we can execute a valid request.

Reply:
oh! Gotcha. For some reason I thought it was logging an event each time a user chose to opt out.

This makes sense though.